octree_web_viewer/src/backend.rs

```diff
@@ -75,106 +74,103 @@ fn pad(input: &mut Vec<u8>) {
-                        })
-                        .unwrap();
-
-                    // Write the bounding box information.
-                    let min = node_data.meta.bounding_cube.min();
-                    reply_blob.write_f64::<LittleEndian>(min.x).unwrap();
-                    reply_blob.write_f64::<LittleEndian>(min.y).unwrap();
-                    reply_blob.write_f64::<LittleEndian>(min.z).unwrap();
-                    reply_blob
-                        .write_f64::<LittleEndian>(node_data.meta.bounding_cube.edge_length())
-                        .unwrap();
-
-                    // Number of points.
-                    reply_blob
-                        .write_u32::<LittleEndian>(node_data.meta.num_points as u32)
-                        .unwrap();
-
-                    // Position encoding.
-                    let bytes_per_coordinate =
-                        node_data.meta.position_encoding.bytes_per_coordinate();
-                    reply_blob.write_u8(bytes_per_coordinate as u8).unwrap();
-                    assert!(
-                        bytes_per_coordinate * node_data.meta.num_points as usize * 3
-                            == node_data.position.len()
-                    );
-                    assert!(node_data.meta.num_points as usize * 3 == node_data.color.len());
-                    pad(&mut reply_blob);
-
-                    reply_blob.append(&mut node_data.position);
-                    pad(&mut reply_blob);
+/// Asynchronous Handler to get Node Data
+/// returns an alias for Box<Future<Item=HttpResponse, Error=Error>>;
+pub fn get_nodes_data(
+    (octree_id, state, nodes): (Path<String>, State<Arc<AppState>>, Json<Vec<String>>),
+) -> FutureResponse<HttpResponse> {
+    let mut start = 0;
+    future::ok(())
+        .and_then(move |_| {
+            start = time::precise_time_ns();
+            let data: Vec<String> = Json::into_inner(nodes);
+            let nodes_to_load = data
+                .into_iter()
+                .map(|e| octree::NodeId::from_str(e.as_str()).unwrap());
+
+            // So this is godawful: We need to get data to the GPU without JavaScript herp-derping with
+            // it - because that will stall interaction. The straight forward approach would be to ship
+            // json with base64 encoded values - unfortunately base64 decoding in JavaScript yields a
+            // string which cannot be used as a buffer. So we would need to manually convert this into
+            // an Array with is very slow.
+            // The alternative is to binary encode the whole request and parse it on the client side,
+            // which requires careful constructing on the server and parsing on the client.
+            let mut reply_blob = Vec::<u8>::new();
+
+            let mut num_nodes_fetched = 0;
+            let mut num_points = 0;
+            let octree: Arc<octree::Octree> =
+                get_octree_from_state(&octree_id.into_inner(), &state).unwrap();
```

Comment:
Can we remove `into_inner()`?

Reply:
this won't compile
`error[E0277]: the trait bound `actix_web::extractor::Path<std::string::String>: std::convert::AsRef<str>` is not satisfied
   --> octree_web_viewer/src/backend.rs:115:17
    |
115 |                 get_octree_from_state(&octree_id, &state).unwrap();
    |                 ^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::AsRef<str>` is not implemented for `actix_web::extractor::Path<std::string::String>``

Reply:
Ah, I see, this is because `octree_id` is not a regular Rust `Path`, but an actix `Path`.